lib/hl7/parser.ex

```diff
@@ -0,0 +1,118 @@
+defmodule HL7.Parser do
+  @carriage_return 4
+  @field 3
+  @repetition 2
+  @component 1
+  @sub_component 0
+
+  def parse(<<"MSH|^~\\&", rest::binary()>> = text) do
+    tokenize(rest, text, 8, 0, ["^~\\&", @field, "|", @field, "MSH"])
+    |> to_lists()
+  end
+
+  defp tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
+    case char do
+      "|" -> tokenize_terminator(rest, original, skip, len, acc, @field)
+      "~" -> tokenize_terminator(rest, original, skip, len, acc, @repetition)
+      "^" -> tokenize_terminator(rest, original, skip, len, acc, @component)
+      "&" -> tokenize_terminator(rest, original, skip, len, acc, @sub_component)
+      "\r" -> tokenize_terminator(rest, original, skip, len, acc, @carriage_return)
+      "\n" -> tokenize_terminator(rest, original, skip, len, acc, @carriage_return)
+      _ -> tokenize(rest, original, skip, len + 1, acc)
+    end
+  end
+
+  defp tokenize("", _original, _skip, 0, acc) do
+    Enum.reverse(acc)
+  end
+
+  defp tokenize("", original, skip, len, acc) do
+    string = binary_part(original, skip, len)
+    Enum.reverse([string | acc])
+  end
+
+  defp tokenize_terminator(text, original, skip, 0, acc, terminator) do
+    tokenize(text, original, skip + 1, 0, [terminator | acc])
+  end
+
+  defp tokenize_terminator(text, original, skip, len, acc, terminator) do
+    string = binary_part(original, skip, len)
+    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc])
+  end
+
+  defp to_lists(tokens) do
+    split_by(tokens, @carriage_return)
+    |> Enum.reject(&(&1 == []))
+    |> Enum.map(&to_segment/1)
+  end
+
+  defp to_segment(tokens) do
```

Comment:
ok, this is brilliant @calvinb ! i _really_ want to put it in... but am torn because it does seem to incur a small perf penalty rather than inlining every option :(

Reply:
I live to slow things down.